### DISCUSSION TITLE
Recommended way to send custom error codes

### DISCUSSION THREAD

**Discussion by mmkal** [trpc/trpc#809] [Q&A] [answered]
We've used `throw new TRPCError(...)` to send back unauthorized and similar standard type responses, but is there a way to send back arbitrary status codes? Use case - two people try to creating conflicting resources, we want to fail the second one with `409 conflict`.

**Comment by KATT** [comment, marked as answer]
The reason why I currently don't allow custom HTTP codes is that I'm trying to be JSON-RPC compatible in the core as I try it to be independent of the transport layer. 

However, I do understand this is might seem a bit too restrictive and I'm open to suggestions on how to make it more flexible. 

Here's how the JSON-RPC codes are defined: https://github.com/trpc/trpc/blob/main/packages/server/src/rpc/codes.ts
And here's the HTTP codes for each are mapped: https://github.com/trpc/trpc/blob/main/packages/server/src/http/internals/getHTTPStatusCode.ts


Ideas welcome!

**Reply by KATT** [reply]
We can also just add your `409` code to the TRPC Error codes, that'd be the quickest fix.

**Reply by KATT** [reply]
We could also probably add a `CUSTOM` type which defaults to `500` but overridable by using [`formatError`](https://trpc.io/docs/error-formatting)'s `httpStatus` (which I'm adding in #811)

**Reply by KATT** [reply]
Even without adding `CUSTOM` you could throw a specific type of error and use `formatError` to specify a custom httpStatus after #811

**Reply by mmkal** [reply]
That last option sounds pretty future proof to me! I'm not familiar with json-rpc but yeah it'd be great to include 409 and 429 (too many requests) as first-class citizens as possible. I haven't poked around in the client much yet but there could be retry logic based on the 4xx response (429 is generally retriable after a delay by well-behaved clients).
